bundles/org.openhab.binding.entsoe/src/main/java/org/openhab/binding/entsoe/internal/EntsoeHandler.java

```diff
@@ -0,0 +1,274 @@
+        }
+    }
+
+    @Override
+    public void dispose() {
+        entsoeTimeSeries.clear();
+        ScheduledFuture<?> refreshJob = this.refreshJob;
+        if (refreshJob != null) {
+            refreshJob.cancel(true);
+            this.refreshJob = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.trace("handleCommand(channelUID:{}, command:{})", channelUID.getAsString(), command.toFullString());
+
+        if (command instanceof RefreshType) {
+            fetchNewPrices();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(EntsoeConfiguration.class);
+        lastDayAheadReceived = ZonedDateTime.of(LocalDateTime.MIN, cetZoneId);
+
+        if (historicDaysInitially == 0) {
+            historicDaysInitially = config.historicDays;
+        }
+
+        if (isLinked(EntsoeBindingConstants.CHANNEL_SPOT_PRICE)) {
+            updateStatus(ThingStatus.UNKNOWN);
+        } else {
+            updateStatus(ThingStatus.ONLINE);
+        }
+
+        refreshJob = scheduler.schedule(this::refreshPrices, 0, TimeUnit.SECONDS);
+    }
+
+    private ZonedDateTime currentCetTime() {
+        return ZonedDateTime.now(cetZoneId);
+    }
+
+    private ZonedDateTime currentCetTimeWholeHours() {
+        return currentCetTime().truncatedTo(ChronoUnit.HOURS);
+    }
+
+    private long getMillisToNextStateUpdate() {
+        Instant now = Instant.now();
+        Instant nextHour = now.truncatedTo(ChronoUnit.HOURS).plus(1, ChronoUnit.HOURS);
+        long millis = Duration.between(now, nextHour).toMillis();
+        try {
+            Instant nextInstant = getActualTimeSerie(true).getInstant();
+            millis = Duration.between(now, nextInstant).toMillis();
+        } catch (EntsoeResponseMapException e) {
+            logger.warn("Using millis to next whole hour");
+        }
+        return millis;
+    }
+
+    private EntsoeTimeSerie getActualTimeSerie() throws EntsoeResponseMapException {
+        return getActualTimeSerie(false);
+    }
+
+    private EntsoeTimeSerie getActualTimeSerie(boolean getNext) throws EntsoeResponseMapException {
+        if (entsoeTimeSeries.isEmpty()) {
+            throw new EntsoeResponseMapException("responseMap is empty");
+        }
+        Duration duration = Duration.parse(config.resolution);
+        for (Map.Entry<Instant, EntsoeTimeSerie> entry : entsoeTimeSeries.entrySet()) {
+            Instant thisEntry = entry.getKey();
+            Instant now = Instant.now();
+            Instant nowPlusNext = now.plus(duration);
+            Instant nowMinusNext = now.minus(duration);
+
+            if (getNext && thisEntry.isAfter(now)) {
+                return entry.getValue();
+            }
+            if (!getNext && thisEntry.isAfter(nowMinusNext) && thisEntry.isBefore(nowPlusNext)) {
+                return entry.getValue();
+            }
+        }
```

Comment:
This seems a bit confusing to me, let me explain:
- The method is doing two different things (parameter `getNext`).
- Should duration really be applied to `now` as a sliding window, or should it actually be applied to the entry to check when now is within the validity period of an entry?
- Start time is **inclusive**, but end time is **exclusive** (just mentioning).
- `now` changes for each loop iteration, so there is a very small risk of not finding the price we are looking for, even if it is there. This can be mitigated by moving lines 159-161 outside the loop.

I'm not sure anything is wrong, but let's see if we can somehow simplify.

Perhaps something like this would be simpler for the `getNext: false` case?
```java
        Duration duration = Duration.parse(config.resolution);
        Instant now = Instant.now();

        for (Map.Entry<Instant, EntsoeTimeSerie> entry : entsoeTimeSeries.entrySet()) {
            Instant entryStart = entry.getKey();
            Instant entryEnd = entryStart.plus(duration);
        
            if (!now.isBefore(entryStart) && now.isBefore(entryEnd)) {
                return entry.getValue();
            }
        }
```

and for the `getNext: true` case:
```java
        Duration duration = Duration.parse(config.resolution);
        Instant now = Instant.now();

        for (Map.Entry<Instant, EntsoeTimeSerie> entry : entsoeTimeSeries.entrySet()) {
            Instant entryStart = entry.getKey();
        
            if (entryStart.isAfter(now)) {
                return entry.getValue();
            }
        }
```

Reply:
ok, extracted into two methods to clarify purpose